### DISCUSSION TITLE
i3 startup: where do I put my call to xrandr?

### DISCUSSION THREAD

**Discussion by jaimet** [i3/i3#5737] [Q&A]
The [`i3` user's guide](https://i3wm.org/docs/userguide.html) includes a [wonderful xrandr tutorial](https://i3wm.org/docs/userguide.html#_configuring_your_monitors), but unfortunately it does not tell me the one thing that I need to know in order to be able to incorporate `xrandr` into my i3 setup, namely where do I put my call to `xrandr`?

I'm running `i3` on Debian 12 (Bookworm) on a laptop with an external monitor attached. I want the laptop's internal screen disabled ("off") and the external monitor as the (only) primary output. I do not use a display manager so I just log-in on vt1/tty1, and then I run `startx`.

If I put the relevant line[^1] into `~/.xsessionrc` and leave `~/.config/i3/config` unchanged, then everything is perfect: the laptop's internal screen is disabled ("off"), the external monitor is the only monitor on my "desktop", and the bar at the bottom of the desktop indicates "workspace 1".

If, however, I add the relevant line[^2] as the _first line_ in `~/.config/i3/config`  (and delete `~/.xsessionrc`) then everything is almost perfect: the laptop's internal screen is disabled ("off"), the external monitor is the only monitor on my "desktop", but the bar at the bottom of the desktop indicates "workspace 2".

AFAICT, this "breakage" is caused because calling `xrandr` in i3's config file is too late - `xrandr` should really be called ASAP after xserver creation and certainly _before_ `i3` starts processing its config file and setting up its outputs.

Can somebody who understands these things please let me know whether my deduction is correct - for `i3` to set up the screen(s) correctly, does `xrandr` need to be called _before_ `i3` starts processing its config file?

TIA.

[^1]: `xrandr --output HDMI-1 --primary --auto --output eDP-1 --off`
[^2]: `exec --no-startup-id xrandr --output HDMI-1 --primary --auto --output eDP-1 --off`

**Comment by Sinaaaa** [comment]
The very first -non commented out- line of my i3 config file is 'exec screensomethingsomething.sh'
The xrandr based display config gets applied without fail. (I have 3 monitors & 2 gpus in a nonstandard array)

**Comment by jaimet** [comment]
> The very first -non commented out- line of my i3 config file is 'exec screensomethingsomething.sh' The xrandr based display config gets applied without fail. (I have 3 monitors & 2 gpus in a nonstandard array)

@Sinaaaa Thank you for your reply.

My problem is _not_ that `xrandr` is is not getting applied, it's how `i3` then assigns workspaces to the remaining output(s).

Does your `screensomethingsomething.sh` disable any of your outputs, or are all your outputs enabled when you start `X`? (I am guessing that I _do_ have this problem and you _don't_ have this problem because my starting setup contains a disabled output whereas your starting setup does _not_ contain a disabled output.)

**Comment by Sinaaaa** [comment]
> (I am guessing that I _do_ have this problem and you _don't_ have this problem because my starting setup contains a disabled output whereas your starting setup does _not_ contain a disabled output.)

I cannot meaningfully help you, but my setup also has a `xrandr` disabled output, I rarely turn on my third display.

**Comment by jaimet** [comment]
> I cannot meaningfully help you, but my setup also has a `xrandr` disabled output, I rarely turn on my third display.

Your comments **have** helped me to understand the problem better - thank you for that.

For anyone who would like to reproduce this "bug", edit your (startup) `xrandr` command to (temporarily) disable your _first_ (inbuilt?) output, then restart `X`. You should find that the workspace allocation is "wrong" (i.e. not as it should be).

I need to do some more research to see how `i3` _is supposed_ to handle workspace (re)allocation due to dynamic `xrandr` changes (while `i3` is running), but as I have now moved back to `bspwm`, that is unfortunately not likely to happen.

I am not going to close this discussion, however, because (in my opinion) it is still unresolved.

**Comment by hoshF** [comment]
I had the same problem, I was booting with sddm, so I put the init file in .xprofile, I wrote my xrandr config in it, but it wouldn't start, in this case I put the xrandr config in i3 config, it would make the mouse spin in circles, it only worked when I manually typed the xrandr command.

**Reply by hoshF** [reply]
I think the key is that i3 uses the X11 protocol, because when I just switched to KDE based on wayland, my screen settings were perfect, different screens used different resolutions, and I felt much smoother, which is a bit of a problem on KDE with X11, you can only adjust the global resolution, no HDR support, etc.
